src/main/java/pl/allegro/tech/opel/OpelEngine.java

```diff
@@ -31,24 +37,43 @@ public ExpressionValidationResult validate(String expression) {
     }
 
     public OpelParsingResult parse(String expression) {
-        return new OpelParsingResult(expression, getParsingResult(expression));
+        return new OpelParsingResult(expression, getParsingResult(expression), embeddedEvalContext);
     }
 
     public CompletableFuture<?> eval(String expression) {
         ParsingResult<ExpressionNode> parsingResult = getParsingResult(expression);
-        return parsingResult.resultValue.getValue();
+        return parsingResult.resultValue.getValue(embeddedEvalContext);
     }
 
     public CompletableFuture<?> eval(String expression, EvalContext evalContext) {
         ParsingResult<ExpressionNode> parsingResult = getParsingResult(expression);
-        return parsingResult.resultValue.getValue(evalContext);
+        return parsingResult.resultValue.getValue(EvalContext.Builder.mergeContexts(evalContext, embeddedEvalContext));
     }
 
     private ParsingResult<ExpressionNode> getParsingResult(String expression) {
         return new ReportingParseRunner<ExpressionNode>(parser.get().ParsingUnit()).run(expression);
     }
 
+    public void registerFunction(String functionName, OpelAsyncFunction<?> function) {
+        embeddedFunctions.put(functionName, function);
+        updateEmbeddedEvalContext();
+    }
+
+    public void registerVariable(String variableName, CompletableFuture<Object> value) {
+        embeddedVariables.put(variableName, value);
+        updateEmbeddedEvalContext();
+    }
+
+    public void registerCompletedVariable(String variableName, Object value) {
+        embeddedVariables.put(variableName, CompletableFuture.completedFuture(value));
+        updateEmbeddedEvalContext();
+    }
+
     public <T, R> void registerImplicitConversion(Class<T> from, Class<R> to, Function<T, R> conversion) {
         implicitConversion.register(new ImplicitConversionUnit<>(from, to, conversion));
     }
+
+    private void updateEmbeddedEvalContext() {
+        embeddedEvalContext = EvalContext.Builder.create().withFunctions(embeddedFunctions).withVariables(embeddedVariables).build();
```

Comment:
I suggest introducing builder together with refactoring of registering implicit conversions so engine is really immutable.